applications/management/utils.py

```diff
@@ -0,0 +1,122 @@
+    """
+    if bootcamp_id1 == bootcamp_id2:
+        return True
+    first_bootcamp_step_types = list(
+        ApplicationStep.objects.filter(bootcamp_id=bootcamp_id1)
+        .order_by("step_order")
+        .values_list("submission_type", flat=True)
+    )
+    second_bootcamp_step_types = list(
+        ApplicationStep.objects.filter(bootcamp_id=bootcamp_id2)
+        .order_by("step_order")
+        .values_list("submission_type", flat=True)
+    )
+    return first_bootcamp_step_types == second_bootcamp_step_types
+
+
+def migrate_application(from_run_application, to_run):
+    """
+    Given an existing application, creates a new application in a different bootcamp run and "migrates" over all of
+    the data from the existing application. Assumes that the 'from' run and 'to' run have the same application steps.
+
+    Args:
+        from_run_application (BootcampApplication): The bootcamp application to copy
+        to_run (BootcampRun): The bootcamp run for which a new application will be created
+
+    Returns:
+        BootcampApplication: The newly-created bootcamp application that was created based on the existing one.
+    """
+    has_completed_app = BootcampApplication.objects.filter(
+        bootcamp_run=to_run,
+        user=from_run_application.user,
+        state__in=APPROVED_APP_STATES,
+    ).exists()
+    if has_completed_app:
+        raise ValidationError(
+            "An approved/completed application already exists for this user and run ({}, {})".format(
+                from_run_application.user.email, to_run.title
+            )
+        )
+
+    with transaction.atomic():
+        to_run_application, _ = BootcampApplication.objects.select_for_update().get_or_create(
+            bootcamp_run=to_run, user=from_run_application.user
+        )
+
+        # Copy work history data
+        if is_empty_file(to_run_application.resume_file) and not is_empty_file(
+            from_run_application.resume_file
+        ):
+            to_run_application.resume_file.name = from_run_application.resume_file.name
+        if (
+            to_run_application.linkedin_url is None
+            and from_run_application.linkedin_url is not None
+        ):
+            to_run_application.linkedin_url = from_run_application.linkedin_url
+        to_run_application.resume_upload_date = from_run_application.resume_upload_date
+        to_run_application.save()
+
+        # Copy submissions (video interview, etc.)
+        from_app_step_submissions = ApplicationStepSubmission.objects.filter(
+            bootcamp_application=from_run_application
+        ).order_by("run_application_step__application_step__step_order")
+        for from_app_step_submission in from_app_step_submissions:
+            ApplicationStepSubmission.objects.update_or_create(
+                bootcamp_application=to_run_application,
+                run_application_step=from_app_step_submission.run_application_step,
```

Comment:
This needs to be linked to the new run's application step instead.  Otherwise the new submission ends up pointing to the original run:

<img width="1242" alt="Screen Shot 2020-11-16 at 2 11 42 PM" src="https://user-images.githubusercontent.com/187676/99297688-963e7800-2816-11eb-8dba-849c151a9364.png">

and although the application has a status of 'AWAITING_PAYMENT', the UI expects a new Video interview:

<img width="963" alt="Screen Shot 2020-11-16 at 2 10 42 PM" src="https://user-images.githubusercontent.com/187676/99297785-b8d09100-2816-11eb-8fef-824238c04b6a.png">

Reply:
Great catch

Reply:
@mbertrand Your comment made me realize that it would be a good idea to handle cases where the 'from' and 'to' bootcamps have repeated steps and/or differently-ordered steps. I added some logic to handle those cases correctly